File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.js
// constructor function
var SmallestInfiniteSet = function() {
    this.popValues = []
};

/**
 * @return {number}
 */
SmallestInfiniteSet.prototype.popSmallest = function() {
    var curLength = this.popValues.length
    
    if (curLength == 0) {
        this.popValues.push(1)
        return 1
    }
    for (var i = 0; i < curLength; i++) {
        var curVal = this.popValues[i]
        var expectedVal = i + 1
        if (curVal != expectedVal) {
            this.popValues = this.popValues.slice(0, i).concat([expectedVal]).concat(this.popValues.slice(i, curLength))
            return expectedVal
        }
    }
    
    this.popValues.push(curLength + 1)    
    return curLength + 1
    
    
    
};

/** 
 * @param {number} num
 * @return {void}
 */
SmallestInfiniteSet.prototype.addBack = function(num) {
    var poppedLen = this.popValues.length
    
    // if poppedValues is empty or the added number has not been popped
    if (poppedLen == 0 || poppedLen[poppedLen - 1] < num || poppedLen[0] > num)
        return
    
    this.popValues = this.popValues.filter(function(item) {
    return item !== num
})
};

/** 
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = new SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
from heapq import *
class SmallestInfiniteSet:

    def __init__(self):
        self.curInteger = 1
        self.addedIntegers = []
        self.addedBack = set()
        

    def popSmallest(self) -> int:
        if self.addedIntegers:
            popped = heappop(self.addedIntegers)
            self.addedBack.remove(popped)
            return popped
        self.curInteger += 1
        return self.curInteger - 1
        
    def addBack(self, num: int) -> None:
        if num >= self.curInteger:
            return
        if num not in self.addedBack:
            self.addedBack.add(num)
            heappush(self.addedIntegers, num)
            return
        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)